File: backend/internal/middleware/auth_rate_limit.go
package middleware

import (
	"time"

	"nexia-backend/internal/config"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
)

const (
	defaultAuthRateLimitRequests = 10
	defaultAuthRateLimitBurst    = 10
	defaultAuthRateLimitWindow   = 10 * time.Second
)

// NewAuthRateLimit returns an in-memory per-IP token-bucket limiter for auth endpoints.
// Requests that exceed the configured budget are rejected with HTTP 429.
func NewAuthRateLimit(logger *zap.Logger, cfg *config.Config) gin.HandlerFunc {
	return newRateLimit(logger, "auth", "Too many authentication requests", authRateLimitConfigFromConfig(cfg))
}

func authRateLimitConfigFromConfig(cfg *config.Config) rateLimitConfig {
	if cfg == nil {
		return rateLimitConfigFromValues(defaultAuthRateLimitRequests, defaultAuthRateLimitBurst, defaultAuthRateLimitWindow, 0, 0, 0)
	}

	return rateLimitConfigFromValues(
		defaultAuthRateLimitRequests,
		defaultAuthRateLimitBurst,
		defaultAuthRateLimitWindow,
		cfg.Server.AuthRateLimitRequests,
		cfg.Server.AuthRateLimitBurst,
		cfg.Server.AuthRateLimitWindowSeconds,
	)
}

File: apps/api/src/middleware/auth-rate-limit.ts
import type { Config } from "../config/config";
import type { Logger } from "pino";
import type { MiddlewareHandler } from "hono";
import { createRateLimiter, rateLimitConfigFromValues, type RateLimitConfig } from "./rate-limit";

const DEFAULT_AUTH_REQUESTS = 10;
const DEFAULT_AUTH_BURST = 10;
const DEFAULT_AUTH_WINDOW_SECONDS = 10;

function authRateLimitConfigFromConfig(cfg: Config | null): RateLimitConfig {
  if (!cfg) {
    return rateLimitConfigFromValues(
      DEFAULT_AUTH_REQUESTS,
      DEFAULT_AUTH_BURST,
      DEFAULT_AUTH_WINDOW_SECONDS,
      0, 0, 0,
    );
  }
  return rateLimitConfigFromValues(
    DEFAULT_AUTH_REQUESTS,
    DEFAULT_AUTH_BURST,
    DEFAULT_AUTH_WINDOW_SECONDS,
    cfg.server.auth_rate_limit_requests,
    cfg.server.auth_rate_limit_burst,
    cfg.server.auth_rate_limit_window_seconds,
  );
}

export function createAuthRateLimiter(logger: Logger, cfg: Config): MiddlewareHandler {
  const rlCfg = authRateLimitConfigFromConfig(cfg);
  return createRateLimiter(logger, "auth", "Too many authentication requests", rlCfg, "/auth");
}
